### PR TITLE
Initial automated tier generation

### PR DESCRIPTION
**Pull request by SUPERCILEX**
@JohnTitor Fixed once and for all! Closes https://github.com/rust-lang/rustup-components-history/pull/40.

The scraping code is a kinda shit, so sorry about that, but the idea is that we want to be able to easily change it. Also of course this is a terrible solution and both pages should be generated from a machine readable format, but I'm guessing it's easier for humans to edit the platform support markdown, so it's not too unreasonable to use that html as the "machine readable" format. Perhaps we can spruce up that page with some divs+ids around the markdown tables to make the scraper more robust.

Finally, I didn't have the energy to write a cron job, but someone should easily be able to run this in GH actions and do auto commits on a change to the config file.

### CODE REVIEW

**Review by JohnTitor — APPROVED** [review]

**Comment by JohnTitor** [comment]
Looks great!
To check the content, could you deploy it on your GitHub Pages?
Also cc @mexus

> Finally, I didn't have the energy to write a cron job, but someone should easily be able to run this in GH actions and do auto commits on a change to the config file.

I can take care of this :)

**Comment by SUPERCILEX** [comment]
Not a single unknown target! :star_struck: https://supercilex.github.io/rustup-components-history/

**Comment by SUPERCILEX** [comment]
@JohnTitor bump

**Comment by JohnTitor** [comment]
The page looks great, thanks a lot!
I don't have access to merge this PR, @Mark-Simulacrum could you take a look when you have time?

File: web/src/main.rs [hunk 1]
```diff
@@ -1,22 +1,24 @@
...
-    fmt::Display,
-    fs::{create_dir_all, File},
-    io::{self, Write},
-    path::{Path, PathBuf},
-};
 use structopt::StructOpt;
 use tiers_table::TiersTable;
 
+mod opts;
+mod tear_ducts;
```
Comment (Mark-Simulacrum) [inline web/src/main.rs:19]: What is the intended meaning of this module name? I don't think we should use a non-descriptive name like this, or at least, it needs a comment.

**Review by Mark-Simulacrum — COMMENTED** [review]
> Finally, I didn't have the energy to write a cron job, but someone should easily be able to run this in GH actions and do auto commits on a change to the config file.

It seems like this repo is already configured with a regular re-run, so getting this to work would be a matter of running the tool to generate configuration prior to running the tool that generates the website?

Am I missing something there?

**Comment by SUPERCILEX** [comment]
> Essentially a one-line change here

Yeah, I guess that could work, but then the build isn't reproducible and will suffer potential reliability issues if the scraper breaks. I was thinking of running a separate config generator job that commits to master when a config change is detected.

File: web/src/main.rs (hunk 1 above)
Comment (SUPERCILEX) [reply web/src/main.rs:19]: > What is the intended meaning of this module name?

Oh lol, the idea is that this is a tier generator and tear ducts create tears/tiers. Happy to change it if that's too convoluted.

Comment (Mark-Simulacrum) [reply web/src/main.rs:19]: Yes, please rename it. Something like config_gen would be not unreasonable I think.

**Comment by Mark-Simulacrum** [comment]
I'm not too worried about it breaking like that. We should dump the config to stdout as part of the build, I think that's enough.

File: web/src/main.rs (hunk 1 above)
Comment (SUPERCILEX) [reply web/src/main.rs]: Done.

**Comment by SUPERCILEX** [comment]
> I'm not too worried about it breaking like that. We should dump the config to stdout as part of the build, I think that's enough.

K, sounds good as long as you're aware of the downsides.

**Comment by SUPERCILEX** [comment]
@Mark-Simulacrum boop

**Comment by SUPERCILEX** [comment]
Thanks everybody!

**Pull request merged**